design/one-pager-component-mutable-and-versioning.md

```diff
@@ -0,0 +1,477 @@
+
+## Example: A blue-green workflow
+
+Let's say we have a `Rollout` Trait with definition like below:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  name: rollouts.extend.oam.dev
+spec:
+  revisionAware: true
+  definitionRef:
+    name: traffics.core.oam.dev
+```
+
+It was used in application with a component.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: frontend
+      traits:
+        - trait:
+            apiVersion: extend.oam.dev/v1
+            kind: Rollout
+            spec:
+              replica: 10
+              batch: 3
+              maxUnavailable: 1
+```
+
+Before changing the component, the App will be running in a stable state like below:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ContainerizedWorkload
+  metadata:
+    name: frontend-v0.0.1
+spec:
+  containers:
+    - name: my-cool-workload
+      image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+      cmd:
+        - "bash lscpu"
+---
+apiVersion: extend.oam.dev/v1
+kind: Rollout
+spec:
+  batch: 3
+  replica: 10
+  maxUnavailable: 1
+  workloadRef:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    name: frontend-v0.0.1
+status:
+  currentWorkloadRef:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    name: frontend-v0.0.1
+```
+
+When `frontend` component changed, new workload will be created with rollout trait updated,
+and the old workload will still be there.
+
+
+```yaml
+# the old one 
+apiVersion: core.oam.dev/v1alpha2
+kind: ContainerizedWorkload
+  metadata:
+    name: frontend-v0.0.1
+spec:
+  containers:
+    - name: my-cool-workload
+      image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+      cmd:
+        - "bash lscpu"
+---
+# the new one 
+apiVersion: core.oam.dev/v1alpha2
+kind: ContainerizedWorkload
+  metadata:
+    name: frontend-v0.0.2
+spec:
+  containers:
+    - name: my-cool-workload
+      image: example/very-cool-workload:new
+      cmd:
+        - "bash top"
+---
+# rollout spec was pointing to the new workload, but status is still refer to the old.
+apiVersion: extend.oam.dev/v1
+kind: Rollout
+spec:
+  batch: 3
+  replica: 10
+  maxUnavailable: 1
+  workloadRef:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    name: frontend-v0.0.2
+status:
+  currentWorkloadRef:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    name: frontend-v0.0.1
+```
+
+Then the rollout trait will take control of the blue-green deploy progress.
+
+Finally the old workload will be removed by rollout trait and the status of rollout trait will align with the spec.
+Getting into another stable state.
+
+
+## Impact to the existing system
+
+1. Add `revisionName` field into AppConfig and this field is mutually exclusive with `componentName`,
```

Comment:
Can we just get away by adding "version" to Component?

Reply:
These two comments above point to the same thing, and it's already discussed here https://github.com/oam-dev/spec/issues/350#issuecomment-617556483.  I'm following the conclusion.